Add actions support to package auth verification

yp05327 opened the pull request:
Partly fixes https://github.com/go-gitea/gitea/issues/23642

Error info:
![image](https://user-images.githubusercontent.com/18380374/227827027-4280a368-ec9e-49e0-bb93-6b496ada7cd9.png)
ActionsUser (userID -2) is used to login in to docker in action jobs.

Due to we have no permission policy settings of ActionsUser now, ActionsUser can only access public registry by this quick fix.

wxiaoguang commented on routers/api/packages/container/auth.go:
```diff
@@ -27,11 +27,16 @@ func (a *Auth) Verify(req *http.Request, w http.ResponseWriter, store auth.DataS
+		switch uid {
+		case -1:
+			return user_model.NewGhostUser(), nil
+		case -2:
```
```suggestion
		case user_model.ActionsUserID:
```

?

IIRC there is a const for it.

yp05327 replied:
I have a question, why `GhostUser` is not defined by consts.

yp05327 commented:
I found that we already have a func called `GetPossibleUserByID` to handle with both normal users and system users.

KN4CK3R reviewed (changes requested):
This will only work if you pull images from a public registry. And you can do that without auth. The action user will not pass permission checks.

wxiaoguang commented:
> This will only work if you pull images from a public registry. And you can do that without auth. The action user will not pass permission checks.

That's also an unclear question in my mind, how the "action" user works with permissions (or how it should work).

Is there any document?

KN4CK3R commented:
open questions I asked in Discord:

> the first question is if that task token should/could be used to access the package registry
if the answer is yes, we need a definition of who is the real user behind that task
is it the owner of the repo, the user which triggered the task?

And after that we may need to change how we use `ctx.Doer`. It's not clear to me where the actions user should be the doer or the real user.

yp05327 commented:
So `GITEA_TOKEN` needs to be `run.TriggerUser`'s temporary token instead of `task.Token`?

yp05327 commented:
`Team.Authorize` and `TeamUnit.AccessMode` confused me.
It seems that we only have `Team.Authorize` before #17811, so maybe `Team.Authorize` is unnecessary now?
If it is unnecessary, maybe we need to do some changes to `GetUserOrgsPermissions` too.
https://github.com/go-gitea/gitea/blob/6706ac2a0f5f2fe4f8e2555be7e2a8b4d5946398/routers/api/v1/org/org.go#L105-L169

wolfogre commented on routers/api/packages/container/container.go:
```diff
@@ -143,6 +144,22 @@ func Authenticate(ctx *context.Context) {
 	if u == nil {
 		u = user_model.NewGhostUser()
+	} else if u.IsActions() {
+		task, err := actions_model.GetTaskByID(ctx, ctx.Data["ActionsTaskID"].(int64))
+		if err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+		}
+		if err := task.LoadJob(ctx); err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+		}
+		if err := task.Job.LoadRun(ctx); err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+		}
+		if err := task.Job.Run.LoadTriggerUser(ctx); err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+		}
```
Is that correct?
If there's an error and the task is nil, and `task.Job` will panic, right?

KN4CK3R commented:
This change needs to work for all package types not just the container registry.

lunny commented on routers/api/packages/api.go:
```diff
@@ -71,6 +72,10 @@ func CommonRoutes(ctx gocontext.Context) *web.Route {
 			return
 		}
+		if err := checkActionUser(ctx); err != nil {
+			ctx.Error(http.StatusInternalServerError, "checkActionUser")
```
`err` doesn't write to log?

wolfogre reviewed (changes requested):
Sorry, I changed my idea. I don't like:

`If user is ActionsUser, it will be changed to Task.Job.Run.TriggerUser to check the permission.`

Maybe it's convenient for users, but it's confusing and I don't see GitHub does in this way. 

It may also lead to permission leakage. Think about this, if I am administrator of gitea.com, and I help with a project hosted on gitea.com (just help, but I don't know/trust the owner). And my PR trigger a workflow, and the operations in it will get my administrator permissions.

I hope we could make it clear:

- If an actions job did something, the doer is the actions bot user.
- If the operation need permissions, we grant the permissions to the actions bot user (could be done in the future).
- If it is inconvenient for you, it's recommended to use access token via secrets.

yp05327 commented:
I agree to add permission check to Action Bot User(maybe we can add an Internal User Type later?)
I will revert the code later.

yardenshoham reviewed (changes requested):
dismiss my review

yp05327 commented:
You are right.
But if we use ActionUser to check permission, contributors can also get some special permissions of the Action Bot User which they don’t have.
This issue can be easily used to do some malicious things, as we can not control who create PRs in sometimes, so all contributors code should be untrusted.
And I think if the admin user want to contribute, they should at least know what the code is doing for, so the risk in this case is less.

wolfogre commented:
> But if we use ActionUser to check permission, contributors can also get some special permissions of the Action Bot User which they don’t have.

https://github.com/go-gitea/gitea/blob/04347eb810689db799003cc342bbbc756716ff12/routers/web/repo/http.go#L210-L220

> This issue can be easily used to do some malicious things, as we can not control who create PRs in sometimes, so all codes from the contributors should be untrusted. It is an unpredictable risk.

https://github.com/go-gitea/gitea/pull/22803

> And I think if the admin user want to contribute, they should at least know what the code is doing for, so the risk in this case is more less. IMO normally admin user is used to manage the site, and not for contributing codes.

It's just a example. What if I am the owner of project A, and contribute to project B? So the workflows in project B will get my permission to project A.

> Maybe we need this:
[docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories)

Yes, see https://github.com/go-gitea/gitea/pull/21937#discussion_r1090229868.

So before this, we need to give permission conservatively. Maybe it is a good choice to keep actions bot user without any permission to packages.

sorenisanerd commented:
This feels scary. I don't think the tokens provided to actions should have the same permissions as any user, regardless of whether they're site admins, project owners, or random bypasses. We should always follow the principle of least privilege.

Each task should get a unique token with permissions as defined in the workflow file or fall back to reasonable defaults.

Github's defaults: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token
Documentation for adjusting the token privileges in actions: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions

yp05327 commented:
> So before this, we need to give permission conservatively. Maybe it is a good choice to keep actions bot user without any permission to packages.

I agree.  
So is it okey to only allow action bot user pull images from public registry in this PR, and add TODO for the permission check of private registry?

wolfogre commented:
> > So before this, we need to give permission conservatively. Maybe it is a good choice to keep actions bot user without any permission to packages.
> 
> I agree.  
> So is it okey to only allow action bot user pull images from public registry in this PR, and add TODO for the permission check of private registry?

👍🏻I believe this is the optimal solution at the moment

yp05327 commented:
Removed permission check in this PR, and added `TODO` in https://github.com/go-gitea/gitea/pull/23879/files#diff-af878eff8f52037e8272d864e1b4d7ace9935f2149f3ed60fcf49cb6be891355R95

wolfogre commented:
> Removed permission check in this PR, and added `TODO` in [#23879 (files)](https://github.com/go-gitea/gitea/pull/23879/files#diff-af878eff8f52037e8272d864e1b4d7ace9935f2149f3ed60fcf49cb6be891355R95)

Please update the description of this PR. 😄

yp05327 commented:
> > Removed permission check in this PR, and added `TODO` in [#23879 (files)](https://github.com/go-gitea/gitea/pull/23879/files#diff-af878eff8f52037e8272d864e1b4d7ace9935f2149f3ed60fcf49cb6be891355R95)
> 
> Please update the description of this PR. 😄

Done.

Maybe it is better to create a summary for this problem? It seems that the discussion is in many different issues/PRs comments.
